[chore] rework of 'method syntax' chapter

remybar opened the pull request:
I've finally reworked the 'method syntax' chapter with small progressive chapters and more examples.

I've also removed the section in 'Traits in Cairo' about using the `generate_trait` attribute when specializing a generic trait.

Tell me if I can still improve/reword some parts.

enitrat commented:
Great overall, however some remarks:
- I would like to see more emphasis on the fact that one of Cairo's restrictions is that type methods don't exist - methods can only be implemented on _traits_ and not on the type directly - unlike rust where you can `impl Type` or OOP where you define a classe's methods inside its body. 
- As such, I think that the Generate Trait section should be moved right after the Defining Method section - where we explain _why_ it's actually useless to define an explicit trait to implement a type's methods if they're not shared between multiple traits

how does that sound to you?

remybar commented:
Yes, I agree ! (even if I don't know Rust 😛 ).

I moved the "generate_trait" chapter just after the "defining methods" one, and then, I updated all the examples to use the `generate_trait` attribute.

Do you think that we should explain more in the "generate trait" chapter why trait definitions are useless in this case ?

enitrat commented:
> Do you think that we should explain more in the "generate trait" chapter why trait definitions are useless in this case ?

if possible, yes

remybar commented on listings/ch04-understanding-ownership/listing_04_02/src/lib.cairo:
done ! I don't know why it has been reformatted 🤔

remybar commented:
> > Do you think that we should explain more in the "generate trait" chapter why trait definitions are useless in this case ?
> 
> if possible, yes. It's not just that the trait definition won't be reused - it's more of a restriciton of the language that doesn't let you define methods on types directly, you always have to go through a trait. But since that trait is actually useless, as in you're only using it to define methods on the type, at least you can abstract it's declaration with this attribute

Ok 👍 
I moved the last paragraph of the introduction which explains the difference between Cairo and Rust, to the beginning of the "The generate_trait attribute" section as it is directly linked to that.
I tried to explain a little bit more but I don't know if it's enough for you. But if you see how to explain it better, don't hesitate to directly modify the text in this PR ;-)

enitrat commented:
i'm using https://www.deepl.com/write/ @remybar

remybar commented:
> i'm using https://www.deepl.com/write/ @remybar

Oh, interesting! I use deepl for its translator tool but I didn't know this feature, thanks 👍
